### ISSUE TITLE
reenable Show Goal id for prooftree

### ISSUE THREAD

**Issue by hendriktews** [coq/coq#10315]
#### Description of the problem

Prooftree (https://askra.de/software/prooftree/), the proof-tree visualization for Coq, relies on the Show Goal "is" command for extracting new subgoals or subgoals that have been changed because of the instantiating of existential variables. Therefore prooftree is broken since Coq 8.7.

#### Feature request: Show Goal <id> at <state>

Could you investigate the feasibility of a command that prints a goal as it looked like in a certain state? The goal id should be the id printed as with coqtop -emacs and the state should be a state number from the coqtop -emacs prompt.

For example, after

    Lemma x: forall(i : nat), exists(j : nat), i = j /\ 1+i = 1+j.
    Proof using.
      eexists.
      split.
        trivial.
      simpl.

the command "Show Goal 13 at 4" should print

    goal ID 13 at state 4 is:

      i : nat
      ============================
      1 + i = 1 + ?j

while "Show Goal 13 at 5" should print

    goal ID 13 at state 5 is:

      i : nat
      ============================
      1 + i = 1 + i

Such a command would make it possible to separate Proof General from prooftree more clearly and to increase performance a little bit by running more proof tree related code in parallel with the next proof command(s). Because, with this feature, the necessary goal printing commands could be inserted into the Coq command queue inside Proof General at an arbitrary later point. Currently, Proof General contains code to decide after each command which goals need to be updated and to inject the appropriate commands before the next proof step.

#### Alternatively re-enable Show Goal <id>

Alternatively, please check if it is possible to enable the Show Goal id command that was present in Coq 8.6 and that apparently went away with PR #765. I apologize for not following the discussion in June 2017 and pointing out that this PR breaks prooftree.

**Comment by jfehrle** [comment]
Hendrik, this looks very routine. It shouldn't be hard to implement.

How helpful is it to have the line "goal ID 13 at state 5 is:" in the output?  (Just thinking out loud.)

It's too late to add this to Coq 8.10, which will probably be completed within a week or so.  8.11 would be about 6 months later.  For you to update your code and to test Coq and prooftree together, I think you'll need to build Coq yourself--unless you can get a compiled version from our continuous integration (CI) process.  Have you built Coq before?  What is your development platform?

Also, do you have time to work on this over the next week or two?  (Much more efficient for both of us that way.)  And if not, when?

@Zimmi48 or @ejgallego Does the CI generate downloadable installation files for each build and platform (preferably for a branch other than master)?  Given that this is a bug fix with a small enhancement, do you think it would be eligible to go into 8.10.1?

> prooftree is broken since Coq 8.7

That's a long while ago, suggesting there are not a lot of users (?).  The Coq documentation should have a section that mentions tools such as prooftree so more people know about it.

The prooftree website says "For Coq >= 8.5, existential variables are severely broken in Prooftree because Coq does not provide enough information, see Coq bug https://github.com/coq/coq/issues/4504."  Is that still an issue?

**Comment by Zimmi48** [comment]
> Given that this is a bug fix with a small enhancement, do you think it would be eligible to go into 8.10.1?

No. That's not a bugfix. Sorry, but this is not a notion that can be twisted as much as one wants.

> The Coq documentation should have a section that mentions tools such as prooftree so more people know about it.

The Coq website has https://coq.inria.fr/related-tools.html. But this is a PG-specific tool, and it is already mentioned in the PG manual: https://proofgeneral.github.io/doc/master/userman/Graphical-Proof_002dTree-Visualization/#Graphical-Proof_002dTree-Visualization.

>Does the CI generate downloadable installation files for each build and platform (preferably for a branch other than master)?

Yes, but on Linux / macOS it's also pretty easy to build Coq from sources, so I wouldn't even bother.

**Comment by jfehrle** [comment]
> No. That's not a bugfix. Sorry, but this is not a notion that can be twisted as much as one wants.

Yes, but is fun to try :-).

**Comment by hendriktews** [comment]
Thanks for your quick answers, I really appreciate your cooperation.

> How helpful is it to have the line "goal ID 13 at state 5 is:" in the output? (Just thinking out loud.)

It is just nice to have but in no way essential. Proof General attaches a retirement action to every command sent to Coq, I can easily store the 13 and the 5 in there.

> It's too late to add this to Coq 8.10, which will probably be

No problem. 8.11 sounds great, I'll work and try on a development branch until then. Compiling myself is no problem.

> Also, do you have time to work on this over the next week or two?

Yes, I am committing to work on this in the next weeks.

> The prooftree website says "For Coq >= 8.5, existential variables are severely broken [...] Is that still an issue?

Yes, I planned to submit a feature request for this later, after there is some progress on the current issue. Do you want to see it now?

> That's a long while ago, suggesting there are not a lot of users (?).

Probably there were no users apart from myself. But I presented it at the 3rd Coq workshop in Nijmegen and promoted it at Tableaux in Nancy. 

> it is already mentioned in the PG manual

Yes, even several times (additionally 1.3 and also 11.7) but it looks like nobody really reads the Proof General manual. Maybe it deserves some better promotion.

**Comment by jfehrle** [comment]
> Do you want to see [broken existential variables] now?

Let's do Show Goal first.

**Comment by jfehrle** [comment]
Hi Hendrik,

The state numbers in your example are off by one.  Please give this a try and let me know if you have any questions/issues.

Sample output using  your example:

```
$ bin/coqtop.exe -coqlib . -emacs <henrik
    <LINES REMOVED>
<prompt>x < 5 |x| 0 < </prompt>1 subgoal (ID 13)

  i : nat
  ============================
  1 + i = 1 + i

<prompt>x < 6 |x| 0 < </prompt>1 subgoal (ID 19)

  i : nat
  ============================
  S i = S i

<prompt>x < 7 |x| 0 < </prompt>goal ID 13 at state 5 is:

  i : nat
  ============================
  1 + i = 1 + ?j

<prompt>x < 8 |x| 0 < </prompt>goal ID 13 at state 6 is:

  i : nat
  ============================
  1 + i = 1 + i
```

**Comment by jfehrle** [comment]
>> That's a long while ago, suggesting there are not a lot of users (?).

>Probably there were no users apart from myself. But I presented it at the 3rd Coq workshop in Nijmegen and promoted it at Tableaux in Nancy. 

>> it is already mentioned in the PG manual

> Yes, even several times (additionally 1.3 and also 11.7) but it looks like nobody really reads the Proof General manual. Maybe it deserves some better promotion.

Once Coq 8.11 is out, I suggest you personally ask several Coq+PG users to try Prooftree and give you candid feedback on it.  Sometimes it takes a little push.  They may have good ideas for improving it, too.  If you get positive feedback, you can ask them to help spread the word to other Coq+PG users.

I think there is a considerable need for tools that help you understand proofs.  Showing proof trees seems like a step in the right direction.

**Comment by hendriktews** [comment]
Thanks for the PR! I will try this out and let you know.

> The state numbers in your example are off by one.

This is probably my fault.

**Comment by hendriktews** [comment]
Jim, the PR is a nice step forward, but I believe, there is still something missing. Let's look at the goal that Coq prints after eexists in lemma x. In your PR, it reads

    <prompt>x < 3 |x| 0 < </prompt>  eexists.
    1 focused subgoal
    (shelved: 1) (ID 10)

      i : nat
      ============================
      i = ?j /\ 1 + i = 1 + ?j

In Coq 8.6 it is the same, only the goal ID is 8 there.

Later in the proof, after trivial instantiated the existential ?j, in Coq 8.6 I can do ``Show Goal "8"'' and get

    goal / evar 8 is:

      i : nat
      ============================
      i = i + 0 /\ 1 + i = 1 + (i + 0)

In your PR I was not able to get this goal after ?j has been instantiated. Show Goal 10 only works at state 4, but not at 6.

Is there a way to also get goal 10 after ?j got instantiated?

**Comment by jfehrle** [comment]
> Is there a way to also get goal 10 after ?j got instantiated?

In a word, no.  Goal 10 only exists in proof state 4.  The behavior of the proof engine may be different now, 4 releases later.  Can you work around this?

I added code to print out valid goal numbers after each tactic.  The code for that goes after the `try` in `Coqloop.top_goal_print`:

```
    begin match newp with
    | None -> ()
    | Some proof ->
      let goals = Proof.all_goals proof in
      Printf.printf "Goals:";
      Goal.Set.iter (fun goal -> Printf.printf " %d" (Evar.repr goal)) goals;
      Printf.printf "\n%!"
    end;
```

And my output is:

```
$ bin/coqtop.exe -coqlib . -emacs <hendrik
Welcome to Coq DESKTOP-MMPNVGH:/cygdrive/c/Users/Jim.DESKTOP-MMPNVGH/Documents/coqSource/coq,unused_nts (9636b8faa0736e404c241acae89b2b2574170668)

<prompt>Coq < 1 || 0 < </prompt>Goals: 6
1 subgoal (ID 6)

  ============================
  forall i : nat, exists j : nat, i = j /\ 1 + i = 1 + j

<prompt>x < 2 |x| 0 < </prompt>Goals: 6

<prompt>x < 3 |x| 0 < </prompt>Goals: 9 10
1 focused subgoal
(shelved: 1) (ID 10)

  i : nat
  ============================
  i = ?j /\ 1 + i = 1 + ?j

<prompt>x < 4 |x| 0 < </prompt>Goals: 9 12 13
2 focused subgoals
(shelved: 1) (ID 12)

  i : nat
  ============================
  i = ?j

subgoal 2 (ID 13) is:
 1 + i = 1 + ?j

<prompt>x < 5 |x| 0 < </prompt>Goals: 13
1 subgoal (ID 13)

  i : nat
  ============================
  1 + i = 1 + i

<prompt>x < 6 |x| 0 < </prompt>Goals: 19
1 subgoal (ID 19)

  i : nat
  ============================
  S i = S i

<prompt>x < 7 |x| 0 < </prompt>goal ID 13 at state 5 is:

  i : nat
  ============================
  1 + i = 1 + ?j
Goals: 19     (* not meaningful *)

<prompt>x < 8 |x| 0 < </prompt>goal ID 13 at state 6 is:

  i : nat
  ============================
  1 + i = 1 + i
Goals: 19     (* not meaningful *)

<prompt>x < 10 |x| 0 < </prompt>
```

**Comment by hendriktews** [comment]
I have worked on proofs where some existential quantifier was eventually instantiated with a big term via several layers of existential variables. In such situations, it is really useful to be able to see a much earlier goal with the current instantiation of existential variables.

For the command ``Show Goal 10 at 6'', would it be possible to take the instantiation of existential variables from state 6 and then search backward to find a state with goal 10 and then print goal 10 from state 4 with the existentials from state 6?

**Comment by jfehrle** [comment]
After Coq executes the `trivial` tactic, there is no longer an existential variable `?j`.  Try experimenting with `Show Existentials.`  Not sure what can be done without changing the proof engine, which is more than I want to take on.

Also note that the printing code is set up to work with a single `Proof.t` object that corresponds to a single proof state.  It is not set up to do substitutions.  While it may be "possible" to do this, it may not be trivial.

**Comment by gares** [comment]
I believe we should clarify that:
- a goal is always implemented by an evar, when it is solved the evar gets an assignment stored in the evar map
- an evar is not necessarily a goal, in particular when a goal is solved its corresponding evar is still in the evar map but is not going to be in the data structures that track which are the open goals

I don't know exactly what "ID" means here (since there is also a notion of evar name that is displayed when the corresponding goal is).

My impression is that @hendriktews would like to be able to print the type of an evar that used to implement goal 10 (I guess ?10) after evar ?j got assigned. The type of ?10 in the evar map at state 6 is the "goal statement with ?j := i". But at state 6 the data structures for goals don't mention goal 10 anymore.

**Comment by gares** [comment]
In other words, the right "key" to access the data is not the "goal" that makes sense only locally, but the evar_key is makes sense for the whole proof forward the evar creation.

**Comment by hendriktews** [comment]
Let me clarify that what I am asking for would actually support two features of prooftree:
1 - drawing a standard prooftree
2 - displaying _one_ goal with different instantiations of existential variables

The current PR would be fully sufficient for feature 1. Therefore, if we come to the conclusion that what we are discussing here is too complicated or requires too much effort, I would still vote for merging the PR.

**Comment by hendriktews** [comment]
> * a goal is always implemented by an evar, when it is solved the evar gets an assignment stored in the evar map

From that I would conclude, that, when an evar is solved, it is only stored in the evar map but other evars or goals that contain the first evar are not updated.

Then I would like to ask: Is it possible to access older states of the evar map? If yes, is it possible to print an evar or goal using an arbitrary evar map?


> I don't know exactly what "ID" means here (since there is also a notion of evar name that is displayed when the corresponding goal is).

This is the printout preceeding the internal goal number when you run coqtop with -emacs. You can see this in my comment from yesterday (June 13). In previous versions of Coq this was the internal evar number corresponding to the goal.

> My impression is that @hendriktews would like to be able to print the type of an evar that used to implement goal 10

Yes, but with an evar map coming from some arbitrary later state and with substituting all evars that have an assignment in the evar map.

> The type of ?10 in the evar map at state 6 is the "goal statement with ?j := i".

Then there is a difference between evar numbers and goal IDs. Lets take the goal after eexists in the example above, which is

    (shelved: 1) (ID 10)

      i : nat
      ============================
      i = ?j /\ 1 + i = 1 + ?j

After trivial, where ?j got assigned in the evar map, I would like
to print this goal again, such that the output is

      i : nat
      ============================
      i = i /\ 1 + i = 1 + i

**Comment by gares** [comment]
If I understand it right, then what you want to print (from the Coq's internal POV) is the type of an evar at different points in the proof. Eg the type of ?10 is the first "sequent"l up there before trivial is run, while it is the second "sequent" if you take the type of ?10 from the evar map after the trivial tactic was run. This is why I suggested to use the evar "id" as the key, and not the goal "id" since goals are more ephemeral.

**Comment by jfehrle** [comment]
Hendrik, It is much better to create a new issue and a new PR for the
evars.  Multiple items in a PR makes the process slower and more
confusing.  More comments later.

**Comment by hendriktews** [comment]
> Hendrik, It is much better to create a new issue and a new PR for the evars.

Fine with me, but I am not sure if there is any action for me now. I believe, the Show Goal ID feature in 8.6 was actually printing evars. Therefore, the new PR for the evars could subsume PR #10356.

**Comment by jfehrle** [comment]
I offered to re-add `Show Goals` and allow specifying the state id.  I'm willing to consider doing the evars if it's also simple, but I don't understand what you want.  I haven't seen a response from you to my comment "After Coq executes the `trivial` tactic, there is no longer an existential variable `?j`. ..." above.  Did you look at `Show Existentials`?  I think you're asking for data the engine doesn't currently provide.  It would be helpful for you to describe exactly what you want (e.g. syntax) for the additional feature with a good example.

The last few comments between you and @gares don't make sense to me.  I'll get into that after I hear back from you.

**Comment by jfehrle** [comment]
Here's the problem:  Before the `trivial`, `Show Existentials` shows

```
Existential 1 = ?j : [i : nat |- nat] (shelved)
Existential 2 = ?Goal : [i : nat |- i = ?j]
Existential 3 = ?Goal0 : [i : nat |- 1 + i = 1 + ?j]
```

After `trivial`, `Show Existentials` shows

```
Existential 1 = ?Goal : [i : nat |- 1 + i = 1 + i]
```

At this point `?j` is no longer present because it has been resolved.

**Comment by hendriktews** [comment]
Jim,

I have the impression, we are reaching the limits of online forum communication. What do you think about a call? What timezone are you in? (I am in UTC+2).

> It would be helpful for you to describe exactly what you want (e.g. syntax) for the additional feature with a good example.

I am asking for two things, where the second is an extension of the first.

1. I am asking for a command that prints a goal at a previous state, with the constraint that the goal was an open goal at this state. Your PR #10356 implements this with Show Goal at `n`. This PR reenables prooftree's basic features (without its features for investigating existential variables). I am sparing the example here, because your PR implements this already, OK?

2. For prooftree's features for existential variables, I am additionally asking for the printing command from point 1, but without the constraint. For instance, after

       Lemma x: forall(i : nat), exists(j k : nat), i = j /\ j = k /\ i = k.
       Proof using.
         eexists.
         eexists.
         split.
           trivial.
         split.
           trivial.

   `Show Goal 16 at 7` should print

       goal ID 16 at state 7 is:

         i : nat
         ============================
         i = i /\ i = ?k /\ i = ?k

   and `Show Goal 16 at 9` should print

       goal ID 16 at state 9 is:

         i : nat
         ============================
         i = i /\ i = i /\ i = i

> After Coq executes the `trivial` tactic, there is no longer an existential variable `?j`. ... above. Did you look at `Show Existentials`?

The documentation says `Show Existentials` only displays all the uninstantiated existential variables, i.e., there might be much more instanciated existentials that are not displayed. Indeed, when you run the above proof script in Coq 8.9.0 with `Set Printing Dependent Evars Line.` as first command, you see

    (dependent evars: ?X12 open, ?X15 open,)

after the second `eexists` command. I believe, 12 and 15 are the internal numbers of the existentials `?j` and `?k`. After the second `trivial`, `Show Existentials` shows no existential variables any more, but there is still

    (dependent evars: ?X12 using , ?X15 using ,)

showing, that the existentials `?j` and `?k` still exist in that state (at least in 8.9.0).

> I think you're asking for data the engine doesn't currently provide.

Maybe - I cannot judge this.

> The last few comments between you and @gares don't make sense to me.
  
Could you point to a particular comment?

**Comment by SkySkimmer** [comment]
Instantiated existentials can be garbage collected, although I'm not sure if it happens without running Optimize Proof.

**Comment by jfehrle** [comment]
@SkySkimmer  I see things like `(dependent evars: ?X12 using , ?X15 using ,)` using 8.9, but when I use master, they are all `(dependent evars:)`.  Should this be considered a regression?

**Comment by jfehrle** [comment]
@gares @SkySkimmer I poked around in `Proof.data.sigma`.  `evar_map` has a field `defn_evars` that seemed promising but when I print it out after the last tactic, the print code doesn't find any names for these variables (hence I show "???") and I don't see anything like `i = i /\ i = ?k /\ i = ?k` in the types.  Nor do I see anything defining the value of `?j` or `?k`.  Unless one of you can tell me how to get the values assigned to `?j` or `?k` I think Hendrik's request is not feasible without changing the engine, which I'm not planning to do.

```
Existential 1 = ??? : [i : nat |- Type]
Existential 2 = ??? : [i : nat  j : nat |- Type]
Existential 3 = ??? : [i : nat  j : nat  k : nat |- Type]
Existential 4 = ??? : [i : nat  j : nat  k : nat |- Type]
Existential 5 = ??? : [i : nat  j : nat  k : nat |- Type]
Existential 6 = ??? : [i : nat  j : nat  k : nat |- Type]
Existential 7 = ??? : [i : nat  j : nat  k : nat |- Type]
Existential 8 = ??? : [i : nat  j : nat  k : nat |- Type]
Existential 9 =
??? : [ |- forall i : nat, exists j k : nat, i = j /\ j = k /\ i = k]
Existential 10 =
??? : [ |- forall i : nat, exists j k : nat, i = j /\ j = k /\ i = k]
Existential 11 = ??? : [i : nat |- exists j k : nat, i = j /\ j = k /\ i = k]
Existential 12 = ??? : [i : nat |- nat]
Existential 13 = ??? : [i : nat |- exists k : nat, i = i /\ i = k /\ i = k]
Existential 14 = ??? : [i : nat |- exists k : nat, i = i /\ i = k /\ i = k]
Existential 15 = ??? : [i : nat |- nat]
Existential 16 = ??? : [i : nat |- i = i /\ i = i /\ i = i]
Existential 17 = ??? : [i : nat |- i = i /\ i = i /\ i = i]
Existential 18 = ??? : [i : nat |- i = i]
Existential 19 = ??? : [i : nat |- i = i /\ i = i]
Existential 20 = ??? : [i : nat |- i = i /\ i = i]
Existential 21 = ??? : [i : nat |- i = i]
```

**Comment by jfehrle** [comment]
@hendriktews Your latest example is clearer and helpful, though the state ID numbers are a little off--every additional command changes the ID numbers.  "Show Goal 16 at 8." gives me the unsubstituted goal:

```
goal ID 16 at state 8 is:

  i : nat
  ============================
  i = ?j /\ ?j = ?k /\ i = ?k
```

> Yes, but with an evar map coming from some arbitrary later state and with substituting all evars that have an assignment in the evar map.

I interpreted "from some arbitrary later state" to mean you wanted something like "Show Goal \<goalID> at \<stateID> substituting \<stateID for the Evar map>", whereas your comments today seem to suggest that you always want to use the evar values from the current proof state.  The earlier comment didn't make sense to me.

Conceptually, combining the substitution into `Show Goal` seems like a non-primitive operation that would only be used by Prooftree.  Adding another command to print the values assigned to defined evars seems more general, perhaps something an average user might use from time to time.  Then Prooftree could do some postprocessing.

Isn't it possible that a single proof could use the same existential variable name `j` for independent evars?  So a textual substitution for `?j` wouldn't always be correct, or would at least require additional attention.

> I have the impression, we are reaching the limits of online forum communication. What do you think about a call? What timezone are you in? (I am in UTC+2).

I think I understand what you want now.  Let's wait to hear back on whether the required info is available from the engine.  We could talk on Skype if you wish.  That would let us step through an example together as I can't exactly reproduce exactly the same goal and state numbers.   I'm in GMT-7.  (FWIW, "UTC" is not a timezone: https://www.timeanddate.com/time/gmt-utc-time.html)

**Comment by hendriktews** [comment]
> @SkySkimmer I see things like `(dependent evars: ?X12 using , ?X15 using ,)` using 8.9, but when I use master, they are all `(dependent evars:)`. Should this be considered a regression?

See also #10399, I consider this a regression.

**Comment by hendriktews** [comment]
> though the state ID numbers are a little off

I am pretty sure the state numbers are correct. I suspect you are not running the script that I posted, for instance because you are running it in Proof General, which silently inserts a lot of commands. Please compare with my terminal output (see below) and post your terminal output, if you are still getting other state numbers.

> "Show Goal 16 at 8." gives me the unsubstituted goal

No. The unsubstituted goal is only printed with state 5.

> The earlier comment didn't make sense to me.

If something does not make sense, please ask.

> I interpreted "from some arbitrary later state" to mean you wanted something like "Show Goal <goalID> at <stateID> substituting <stateID for the Evar map>",

Please run the script that I posted yesterday in a fresh coqtop, without adding additional commands. After the last `trivial` coqtop reaches state 9. In this state `Show Goal 16 at 7` should print, what I indicated yesterday. There are at least 3 states involved here:
1. state 5, in which goal 16 was valid
2. state 7, from which I would like the get the evar map
3. state 9, at which the command is issued.

> whereas your comments today seem to suggest that you always want to use the evar values from the current proof state.

No. Yesterday I wrote "after [the script] `Show Goal 16 at 7` should print ... and `Show Goal 16 at 9` should print ...". Because the printout is different, there must two different evar maps be involved.

> Conceptually, combining the substitution into `Show Goal` seems like a non-primitive operation that would only be used by Prooftree.

I am a bit surprised. In the example from yesterday, after the first `split`, the second subgoal is printed as

    subgoal 2 (ID 19) is:
     ?j = ?k /\ i = ?k

If you then (after the first `split`) do `instantiate (1 := i).`, the second subgoal is printed as

    subgoal 2 (ID 19) is:
     i = ?k /\ i = ?k

Doesn't this indicate that printing a goal with a different evar map is a primitive operation, which is used quite often?

> Then Prooftree could do some postprocessing.

Yes, but this would require an almost full parser and pretty printer for terms in prooftree.

> Isn't it possible that a single proof could use the same existential variable name `j` for independent evars? So a textual substitution for `?j` wouldn't always be correct, or would at least require additional attention.

Yes, with `Lemma x: forall(i : nat), exists(j : nat), i = j /\ exists(j : nat), i = j.` and just using eexists, split and trivial, one gets `?j` both for the first `j` and the second `j`.

Here is my terminal output for the example from yesterday.

    3 cert ~ 4> ledit src/orig/coq/install-pr10356/bin/coqtop -emacs
    Welcome to Coq cert:/home/tews/src/orig/coq/coq-repo,pr/10356 (cdda053d412d789518a815761704faa74c911a85)

    <prompt>Coq < 1 || 0 < </prompt>Lemma x: forall(i : nat), exists(j k : nat), i = j /\ j = k /\ i = k.
    1 subgoal (ID 9)

      ============================
      forall i : nat, exists j k : nat, i = j /\ j = k /\ i = k

    <prompt>x < 2 |x| 0 < </prompt>Proof using.

    <prompt>x < 3 |x| 0 < </prompt>  eexists.
    1 focused subgoal
    (shelved: 1) (ID 13)

      i : nat
      ============================
      exists k : nat, i = ?j /\ ?j = k /\ i = k

    <prompt>x < 4 |x| 0 < </prompt>  eexists.
    1 focused subgoal
    (shelved: 2) (ID 16)

      i : nat
      ============================
      i = ?j /\ ?j = ?k /\ i = ?k

    <prompt>x < 5 |x| 0 < </prompt>  split.
    2 focused subgoals
    (shelved: 2) (ID 18)

      i : nat
      ============================
      i = ?j

    subgoal 2 (ID 19) is:
     ?j = ?k /\ i = ?k

    <prompt>x < 6 |x| 0 < </prompt>    trivial.
    1 focused subgoal
    (shelved: 1) (ID 19)

      i : nat
      ============================
      i = ?k /\ i = ?k

    <prompt>x < 7 |x| 0 < </prompt>  split.
    2 focused subgoals
    (shelved: 1) (ID 21)

      i : nat
      ============================
      i = ?k

    subgoal 2 (ID 22) is:
     i = ?k

    <prompt>x < 8 |x| 0 < </prompt>    trivial.
    1 subgoal (ID 22)

      i : nat
      ============================
      i = i

    <prompt>x < 9 |x| 0 < </prompt>Show Goal 16 at 5.
    goal ID 16 at state 5 is:

      i : nat
      ============================
      i = ?j /\ ?j = ?k /\ i = ?k

    <prompt>x < 10 |x| 0 < </prompt>Show Goal 16 at 7.
    Toplevel input, characters 0-18:
    > Show Goal 16 at 7.
    > ^^^^^^^^^^^^^^^^^^
    Error: No such goal.

    <prompt>x < 10 |x| 0 < </prompt>Show Goal 16 at 8.
    Toplevel input, characters 0-18:
    > Show Goal 16 at 8.
    > ^^^^^^^^^^^^^^^^^^
    Error: No such goal.

**Comment by jfehrle** [comment]
OK, I replicated that output.

"Show Goal 16 at 7" and "Show Goal 16 at 8" fail in your output because goal 16 isn't defined in those states.  It would have been much clearer if you had given both the exact current output and the exact desired output in yesterday's example.  I thought you were asking only to change what's printed when the goal exists.

Regarding goal ID 19 appearing in 2 different forms in different states:  When a goal is changed in a proof step, it's usually assigned a new goal ID.  This seems to be a case where that convention is not followed, rather than printing with a different map.

I think further discussion doesn't matter unless someone more expert than me can explain how to get the evar definitions from the engine.  And this has already more time than I planned to spend on it.

**Comment by ejgallego** [comment]
I am getting a bit concerned, it seems my comments on re-adding the wrong command are not being taken into account.

**Comment by jfehrle** [comment]
Ah, my impression from https://github.com/coq/coq/issues/10315#issuecomment-503266662 was that item 2 was new functionality rather than the original behavior so I didn't understand the point of your comment.  My bad.  (Still working on the test case.)

Hendrik, would you try this updated version?  If this is good, what is the (I thought separate) evar issue you have?

**Comment by ejgallego** [comment]
@jfehrle I think @hendriktews is asking for the original functionality, however the code you are adding in #10315 is not the original one. I am not sure what is going on here but I am a bit lost on why you just didn't resurrect the original `evar` printing code.

**Comment by jfehrle** [comment]
I added the new state id ("sid") parameter that Hendrik asked for and as a result moved the command to toplevel, so the code won't be identical to the old code.

**Comment by ejgallego** [comment]
I am talking about the functionality. Have you read the previous implementation?

**Comment by hendriktews** [comment]
> was that item 2 was new functionality rather than the original behavior

Yes and no. The original version did not have the state parameter, but was able to print goals that did not exist in the current state. Going back to the example from yesterday, in order to print goal 16 with the evar instantiation at state 7, one would have had to do `Show Goal "16"` at state 7, that is, directly after the first `trivial`.

I asked for the state parameter on top of the original feature, because that would simplify the prooftree code.

**Comment by hendriktews** [comment]
> @jfehrle I think @hendriktews is asking for the original functionality

As I just wrote, I asked for the state parameter in addition to the original functionality. This way I can do `Show Goal 16 at 7` in an arbitrary state greater or equal to 7. With this I can move prooftree code out of the inner Proof General command loop into the external prooftree program, making it more efficient (at least theoretically).

**Comment by jfehrle** [comment]
> Have you read the previous implementation?

Yes.  I incorporated a key change in the last push. I'm now using (from the old version)

```
      Proof.in_proof proof (fun sigma -> pr {it=(Evar.unsafe_of_int gid);sigma=sigma;})
```

in Printer.pr_goal_emacs whereas before it was 

```
      let goal = Goal.Set.find (Evar.unsafe_of_int gid) goals in
      Proof.in_proof proof (fun sigma ->
          Pp.(pr_selected_subgoal (Some (str "goal ID " ++ (int gid) ++ str " at state " ++ (int sid)))
            (str (string_of_int gid)) sigma goal))
```
```

**Comment by jfehrle** [comment]
@hendriktews Here's the output I get now for yesterday's example.  Does this address your needs?  Should we commit the change?

```
<prompt>x < 9 |x| 0 < </prompt>goal ID 16 at state 5

  i : nat
  ============================
  i = ?j /\ ?j = ?k /\ i = ?k

<prompt>x < 9 |x| 0 < </prompt>goal ID 16 at state 7

  i : nat
  ============================
  i = i /\ i = ?k /\ i = ?k

<prompt>x < 9 |x| 0 < </prompt>goal ID 16 at state 9

  i : nat
  ============================
  i = i /\ i = i /\ i = i
```

**Comment by hendriktews** [comment]
> Does this address your needs? Should we commit the change?

Yes and yes. See my comments on PR #10356. Thanks again!

**Comment by hendriktews** [comment]
> what is the (I thought separate) evar issue you have?

Please see #10420.

**Comment by jfehrle** [comment]
https://github.com/coq/coq/pull/10412 added the output-coqtop test directory.  I moved the test there and it works as expected.  Hendrik says this addresses his needs.  @ejgallego would you merge now?  Thanks.
